### PR TITLE
Feature/add self supervision remaining

### PR DESCRIPTION
**Pull request by eduardocarvp**
- [ ] 

**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

*Any change needs to be discussed before proceeding. Failure to do so may result in the rejection of the pull request.*

**What kind of change does this PR introduce?**

**Does this PR introduce a breaking change?**

**What needs to be documented once your changes are merged?**

**Closing issues**

Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if such).

### CODE REVIEW

File: pytorch_tabnet/metrics.py [hunk 1]
```diff
@@ -24,20 +24,26 @@ def UnsupervisedLoss(y_pred, embedded_x, obf_vars, eps=1e-9):
...
     Returns
     -------
     loss : torch float
-        Unsupervised loss, average value over batch samples.    
+        Unsupervised loss, average value over batch samples.
     """
     errors = y_pred - embedded_x
-    reconstruction_errors = torch.mul(errors, obf_vars)**2
-    batch_stds = torch.std(embedded_x, dim=0)**2 + eps
+    reconstruction_errors = torch.mul(errors, obf_vars) ** 2
```
Comment (mdeboc) [inline pytorch_tabnet/metrics.py:38]: We may call this `masked_vars`  instead of `obf_vars`

File: pytorch_tabnet/metrics.py [hunk 2]
```diff
@@ -24,20 +24,26 @@ def UnsupervisedLoss(y_pred, embedded_x, obf_vars, eps=1e-9):
...
+    batch_stds = torch.std(embedded_x, dim=0) ** 2 + eps
     features_loss = torch.matmul(reconstruction_errors, 1 / batch_stds)
+    # compute the number of non-obfuscated variables
+    nb_used_variables = torch.sum(1 - obf_vars, dim=1)
```
Comment (mdeboc) [inline pytorch_tabnet/metrics.py:42]: better have:
```python
    nb_masked_variables = torch.sum(masked_vars, dim=1)
```
than
```python
    nb_used_variables = torch.sum(1 - obf_vars, dim=1)
```

File: pytorch_tabnet/metrics.py [hunk 3]
```diff
@@ -24,20 +24,26 @@ def UnsupervisedLoss(y_pred, embedded_x, obf_vars, eps=1e-9):
...
+    # print(nb_used_variables)
+    # take the mean of the used variable errors
+    # print(features_loss)
+    features_loss = features_loss / (1 + nb_used_variables)
```
Comment (mdeboc) [inline pytorch_tabnet/metrics.py:46]:
```python
    features_loss = features_loss / (1 + nb_masked_variables)
```

**Review by mdeboc — CHANGES_REQUESTED** [review]

File: pytorch_tabnet/metrics.py (hunk 1 above)
Comment (Optimox) [reply pytorch_tabnet/metrics.py:38]: `obf_vars` stands for obfuscated variables, I tried to make the distinction between obfuscated variables for self supervised training and masked variables which are masked by the attention layers. I think I prefer `obf_vars` but maybe it could be renamed `obfuscated_vars` for clarity.

File: pytorch_tabnet/metrics.py (hunk 2 above)
Comment (Optimox) [reply pytorch_tabnet/metrics.py:42]: well here we need the number of used variables and not the number of obfuscated variables so I'm not sure to understand your point?

File: pytorch_tabnet/metrics.py (hunk 3 above)
Comment (Optimox) [reply pytorch_tabnet/metrics.py:46]: why is there a 1 here? shouldn't this be an epsilon? So that we avoid 0 division error but don't change the results when there are masked_variables?

File: pytorch_tabnet/metrics.py (hunk 2 above)
Comment (mdeboc) [reply pytorch_tabnet/metrics.py:42]: Indeed, that's what I mean, based on my understanding of the situation.
But maybe I am wrong. If so, a comment might explain somewhere why we need to 'correct' the `features_loss` when all variables are used (none hidden)

File: pretraining_example.ipynb [hunk 4]
```diff
@@ -202,15 +202,15 @@
...
+    "scrolled": false
    },
    "outputs": [],
    "source": [
     "unsupervised_model.fit(\n",
     "    X_train=X_train,\n",
     "    eval_set=[X_valid],\n",
-    "    max_epochs=max_epochs , patience=5,\n",
-    "    batch_size=1024, virtual_batch_size=128,\n",
+    "    max_epochs=1 , patience=5,\n",
```
Comment (Optimox) [inline pretraining_example.ipynb:212]: I guess we should switch back to max_epochs > 1

File: pytorch_tabnet/metrics.py (hunk 2 above)
Comment (Optimox) [reply pytorch_tabnet/metrics.py:42]: Yes I think what we want to do is normalize the loss so that the range is not dependent on the number of reconstructed features. so we should indeed compute `nb_reconstructed_features = torch.sum(obf_vars, dim=1)`

File: pytorch_tabnet/metrics.py (hunk 3 above)
Comment (Optimox) [reply pytorch_tabnet/metrics.py:46]: I still think that we should divide by `epsilon + nb_reconstructed_features` it's safe to do so since when `nb_reconstructed_features=0` then the numerator should be zero as well. So the loss won't blow up.

**Review by Optimox — CHANGES_REQUESTED** [review]

File: pytorch_tabnet/metrics.py (hunk 2 above)
Comment (eduardocarvp) [reply pytorch_tabnet/metrics.py]: You're right, I mixed up the masks' meaning.
